File: 0003_longest_substring_without_repeating_characters/0003.py
#!/usr/bin/env python3
# -*-coding:utf-8 -*-
# =============================================================================
"""
@Probelm :   0003.py (Medium)
@First   :   2022/04/15
@URL     :   https://leetcode.com/problems/longest-substring-without-repeating-characters/
"""
# =============================================================================

# from heapq import *
# from typing import List, Optional

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        small, big = 0, 0
        res = 0
        char = dict()
        while big < len(s):
            if s[big] in char:
                # find the next position of the same char
                small = max(small, char[s[big]]+1)
            # update char-index mapping
            char[s[big]] = big
            # update result
            res = max(res, big-small+1)
            big += 1
        return res


def main():
    s = Solution()
    ss = "pwwkew"
    # 3
    print(s.lengthOfLongestSubstring(ss))

if __name__ == "__main__":
    main()

File: 0003_longest_substring_without_repeating_characters/0003.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int small = 0;
        int big = 0;
        unordered_map<int, int> mapping;
        int res = 0;
        while (big < s.length()){
            if (mapping.count(s[big])){
                small = max(small, mapping[s[big]]+1);
            }
            mapping[s[big]] = big;
            res = max(res, big - small + 1);
            big ++;
        }
        return res;
    }
};
